ykrt/src/compile/jitc_yk/codegen/x64/lsregalloc.rs

```diff
@@ -1814,6 +1764,93 @@ pub(crate) enum GPConstraint {
     None,
 }
 
+/// Takes a map of "register X should copy the value of register Y" and turn it into a sequence of
+/// copies respecting dependencies.
+///
+/// For example, if we say that "R8 should copy R9 and R7 should copy R8" we must do the second
+/// copy first to avoid R7 ending up with the value originally in R9. Furthermore, there can be
+/// (direct or indirect) cycles in the input such as "R8 should copy R9 and R9 should copy R8":
+/// this function will first copy one of those registers into a temporary register, and then
+/// generate two copies, one of which will copy from that temporary register.
+///
+/// The input format is: one copy per `Some`. For example `[None, Some(R15), Some(RDX)...]` means
+/// "RAX doesn't copy anything; RCX copies R15's value; and RDX's value remains unchanged". When a
+/// register's value remains unchanged a `RegAction::Keep` is generated.
+fn reg_copies_to_actions(moves: [Option<Rq>; 16]) -> Vec<(Rq, RegAction)> {
+    let mut actions = Vec::new();
+    let mut action_regset = RegSet::with_gp_reserved();
+
+    // Step 1: get the actions into a dependency order.
+    for (to_reg_i, from_reg) in moves.into_iter().enumerate() {
+        if let Some(from_reg) = from_reg {
+            let to_reg = GP_REGS[to_reg_i];
+            action_regset.set(to_reg);
+            if from_reg == to_reg {
+                // The operand is already in the correct register.
+                actions.push((to_reg, RegAction::Keep));
+                continue;
+            }
+            match actions
+                .iter()
+                .position(|(x, action)| *x == from_reg && matches!(action, RegAction::CopyFrom(_)))
+            {
+                Some(i) => {
+                    actions.insert(i, (to_reg, RegAction::CopyFrom(from_reg)));
+                }
+                None => {
+                    actions.push((to_reg, RegAction::CopyFrom(from_reg)));
+                }
+            }
+        }
+    }
+
+    // Step 2: Deal with cycles
```

Comment:
Could the `xchg` instruction be useful here? Maybe not now, but in the future?

Reply:
`xchg` is very slow because it is also (weirdly IMHO) a memory barrier. Using a temporary register is much faster.